### PR TITLE
Adding ESMF.LocStream capabilities to xESMF

### PR DESCRIPTION
**Pull request by raphaeldussin**
@JiaweiZhuang this PR allows xESMF to use LocStream objects (sequence of geographical points).
This is particularly useful to interpolate onto ocean sections, creating open boundary conditions,...

I have made minimal inserts, that do not change any default behavior. Also added some testing.
Let me know if there's more to be done for the PR.

Best,
Raf

### CODE REVIEW

File: xesmf/frontend.py [hunk 1]
```diff
@@ -72,9 +72,45 @@ def ds_to_ESMFgrid(ds, need_bounds=False, periodic=None, append=None):
...
+    locstream = esmf_locstream(lon, lat)
+
+    return locstream, (1,) + lon.shape
+
+
 class Regridder(object):
     def __init__(self, ds_in, ds_out, method, periodic=False,
-                 filename=None, reuse_weights=False, ignore_degenerate=None):
+                 filename=None, reuse_weights=False, ignore_degenerate=None,
+                 locstream_out=False):
```
Comment (JiaweiZhuang) [inline xesmf/frontend.py]: `locstream` might also be input I think? (regridding locstream -> grid)

File: xesmf/frontend.py [hunk 2]
```diff
@@ -72,9 +72,45 @@ def ds_to_ESMFgrid(ds, need_bounds=False, periodic=None, append=None):
...
+
+    if len(lon.shape) > 2:
+        raise ValueError("lon can only be 1d or 2d")
+    if len(lat.shape) > 2:
+        raise ValueError("lat can only be 1d or 2d")
+
+    if len(lon.shape) == 2:
+        lon = lon.flatten()
+    if len(lat.shape) == 2:
+        lat = lat.flatten()
```
Comment (JiaweiZhuang) [inline xesmf/frontend.py]: Better require strict 1-D shape which is the definition of locstream.

**Comment by JiaweiZhuang** [comment]
Thanks very much for the PR! This is something that I've been wanting to add but never got time to actually do it. Also thanks for the unit test. What would be even more useful is an example notebook under [doc/notebooks](https://github.com/JiaweiZhuang/xESMF/tree/master/doc/notebooks) to show a use case.

I am mostly thinking about how to have a consistent API to handle various cases:
- grid -> grid (current only option)
- grid -> locstream (your PR)
- locstream -> grid (this may have more demand, as the above one can also be achieved by xarray [Advanced Interpolation](http://xarray.pydata.org/en/stable/interpolation.html#advanced-interpolation))
- locstream -> locstream (also possible but rarely used)

Options are:
1. Add `locstream_out` and `locstream_in` kwarg to `Regridder`, following your PR
2. Have a new subclass `LocStreamRegridder` to handle locstream input/output, and avoid touching the original `Regridder` API. It might feel like [different classes under sklearn linear regressors](https://scikit-learn.org/stable/modules/classes.html#classical-linear-regressors)
3. Mark the locstream attribute in the grid object `ds_in` / `ds_out` (detect coordinate name like `locations`, but it won't work for dict), and avoid additional kwargs or classes.

I was leaning towards no. 2, but your approach also makes sense.

**Comment by JiaweiZhuang** [comment]
Another complication from LocStream is that this class can only be used for
- Destination for bilinear and patch (cannot be source)
- Source and destination for nearest neighbor 
It cannot be used for conservative method, for either source or destination.

Ref: [ESMF Regridding Update](https://www.earthsystemcog.org/site_media/projects/esmf/pres_0415_cw2015_gave_forweb.ppt)

File: xesmf/tests/test_frontend.py [hunk 3]
```diff
@@ -46,8 +50,14 @@ def test_as_2d_mesh():
 methods_list = ['bilinear', 'conservative', 'nearest_s2d', 'nearest_d2s']
 
 @pytest.mark.parametrize('method', methods_list)
-def test_build_regridder(method):
-    regridder = xe.Regridder(ds_in, ds_out, method)
+@pytest.mark.parametrize('locstream', [True, False])
+def test_build_regridder(method, locstream):
```
Comment (JiaweiZhuang) [inline xesmf/tests/test_frontend.py]: I don't think `locstream` can be used with `conservative` method. Does it actually work here?

File: xesmf/frontend.py (hunk 2 above)
Comment (raphaeldussin) [reply xesmf/frontend.py]: I'm fine with it! maybe the flexibility is not a good thing here actually.

File: xesmf/tests/test_frontend.py (hunk 3 above)
Comment (raphaeldussin) [reply xesmf/tests/test_frontend.py]: no I hacked it on line 57 :)

**Comment by raphaeldussin** [comment]
About the more general comments:

* as an oceanographer, I wouldn't use locstream -> grid when there's more powerful tools such as DIVA http://modb.oce.ulg.ac.be/mediawiki/index.php/DIVA

* locstream to locstream also looks very exotic

I went for option 1 because that's in my opinion, the one that added the less code. Once the ESMF Fields are created, ESMF Regridder works the same (not allowing all types of regridding of course).
I think adding a locstream_in a way that's symetric to locstream_out should not be a big issues if you
think that might be useful (see previous point). Even with LocStream in, I think 95+% of users will be coming for the grid-> grid interpolation. What do you think?

I'm gonna put some more tests, the codecov for my PR is not very good. An example notebook with follow once we ironed out the API details.

File: xesmf/tests/test_frontend.py (hunk 3 above)
Comment (JiaweiZhuang) [reply xesmf/tests/test_frontend.py]: I would use more explicit expression and avoid overwriting `method` internally:

```python
@pytest.mark.parametrize(
    "locstream,method", [
     (False, method_list), 
     (True, ['bilinear', 'nearest_s2d', 'nearest_d2s'])   # conservative method does not work for LocStream
])
```

**Comment by JiaweiZhuang** [comment]
> I think adding a locstream_in a way that's symetric to locstream_out should not be a big issues if you think that might be useful

Yeah I think it is good to have both `locstream_out` and  `locstream_in`  as options. The latter is useful for aligning observation data to model grid (similar to [supersampling](https://en.wikipedia.org/wiki/Supersampling)).

For the example notebook, it can be with synthetic data replicating [advanced interpolation](http://xarray.pydata.org/en/stable/interpolation.html#advanced-interpolation) or with real data from `xr.tutorial`. Can revert the resulting locstream back to model grid via `locstream_in`.

**Comment by raphaeldussin** [comment]
@JiaweiZhuang sorry for the repeated commits, it's ready to be reviewed/merged.

**Comment by raphaeldussin** [comment]
@JiaweiZhuang I realized that I have missed one of your comments. Fixed now and I also expanded the test_build_regridder to include locstream_in.

Let me know if there's more that needs to be done. I'd really like to move fwd with it.

File: xesmf/tests/test_frontend.py [hunk 4]
```diff
@@ -250,6 +250,7 @@ def test_regrid_dask():
     # clean-up
     regridder.clean_weight_file()
 
+    # -----------------------------------------
```
Comment (JiaweiZhuang) [inline xesmf/tests/test_frontend.py]: I would separate out of the tests for locstreams into different functions

File: xesmf/tests/test_frontend.py [hunk 5]
```diff
@@ -259,6 +260,15 @@ def test_regrid_dask():
...
+    # testing locstream in
+    regridder = xe.Regridder(ds_locs, ds_in, 'nearest_s2d', locstream_in=True)
+
+    outdata = regridder(ds_locs['lat'].data) 
+
+    # clean-up
+    regridder.clean_weight_file()
+
```
Comment (JiaweiZhuang) [inline xesmf/tests/test_frontend.py]: I would separate out of the tests for locstreams into different functions

File: xesmf/tests/test_frontend.py [hunk 6]
```diff
@@ -293,6 +303,15 @@ def test_regrid_dataarray_dask():
...
 
+    # -----------------------------------------
+    # testing locstream in
+    regridder = xe.Regridder(ds_locs, ds_in, 'nearest_s2d', locstream_in=True)
+
+    outdata = regridder(ds_locs['lat']) 
+
+    # clean-up
+    regridder.clean_weight_file()
+
```
Comment (JiaweiZhuang) [inline xesmf/tests/test_frontend.py]: (same as above) I would separate out of the tests for locstreams into different functions

File: xesmf/tests/test_frontend.py [hunk 7]
```diff
@@ -323,12 +342,22 @@ def test_regrid_dataset():
...
 
+    # -----------------------------------------
+    # testing locstream in
+    regridder = xe.Regridder(ds_locs, ds_in, 'nearest_s2d', locstream_in=True)
+
+    outdata = regridder(ds_locs) 
+
+    # clean-up
+    regridder.clean_weight_file()
+
```
Comment (JiaweiZhuang) [inline xesmf/tests/test_frontend.py]: (same as above) I would separate out of the tests for locstreams into different functions

**Comment by JiaweiZhuang** [comment]
@raphaeldussin Sorry I had a long backlog. A few more comments on the test code, and it should be good to go.

File: xesmf/tests/test_frontend.py [hunk 8]
```diff
@@ -260,3 +352,32 @@ def test_regrid_dataset():
...
+    # -----------------------------------------
+    # testing locstream in
+    regridder = xe.Regridder(ds_locs, ds_in, 'nearest_s2d', locstream_in=True)
+
+    outdata = regridder(ds_locs) 
+
+    # clean-up
+    regridder.clean_weight_file()
```
Comment (JiaweiZhuang) [inline xesmf/tests/test_frontend.py]: I would separate out the tests for locstreams into different functions

File: xesmf/tests/test_frontend.py [hunk 9]
```diff
@@ -230,6 +305,23 @@ def test_regrid_dataarray_dask():
...
+
+    # -----------------------------------------
+    # testing locstream in
+    regridder = xe.Regridder(ds_locs, ds_in, 'nearest_s2d', locstream_in=True)
+
+    outdata = regridder(ds_locs['lat']) 
+
+    # clean-up
```
Comment (JiaweiZhuang) [inline xesmf/tests/test_frontend.py]: I would separate out the tests for locstreams into different functions

File: xesmf/tests/test_frontend.py [hunk 10]
```diff
@@ -204,6 +261,24 @@ def test_regrid_dask():
...
+    regridder.clean_weight_file()
+
+    # -----------------------------------------
+    # testing locstream in
+    regridder = xe.Regridder(ds_locs, ds_in, 'nearest_s2d', locstream_in=True)
+
+    outdata = regridder(ds_locs['lat'].data) 
+
+    # clean-up
+    regridder.clean_weight_file()
```
Comment (JiaweiZhuang) [inline xesmf/tests/test_frontend.py]: I would separate out the tests for locstreams into different functions

File: xesmf/tests/test_frontend.py [hunk 11]
```diff
@@ -184,6 +206,41 @@ def test_regrid_dataarray():
...
+    regridder = xe.Regridder(ds_locs, ds_in, 'nearest_s2d', locstream_in=True)
+
+    outdata = regridder(ds_locs['lat'].values)  # pure numpy array
+    dr_out = regridder(ds_locs['lat'])  # xarray DataArray
+
+    # DataArray and numpy array should lead to the same result
+    assert_equal(outdata, dr_out.values)
+
+    # clean-up
+    regridder.clean_weight_file()
```
Comment (JiaweiZhuang) [inline xesmf/tests/test_frontend.py]: I would separate out the tests for locstreams into different functions

**Comment by raphaeldussin** [comment]
move the locstream tests into their own functions.

**Comment by JiaweiZhuang** [comment]
Thanks @raphaeldussin! Added to 0.3.0 release.

**Pull request merged**